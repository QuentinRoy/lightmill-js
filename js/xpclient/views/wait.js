/*jslint nomen: true, browser:true*/
/*global define */

define(['spin', 'jquery'], function (Spinner, $) {
    "use strict";

    function WaitView(parentView) {
        this._parentView = parentView;
        this._spinView = $('<div class="full-parent"></div>');
        this._spinner = new Spinner(this._opts);
    }

    WaitView.prototype = {

        _opts: {
            lines: 11, // The number of lines to draw
            length: 20, // The length of each line
            width: 10, // The line thickness
            radius: 30, // The radius of the inner circle
            corners: 1, // Corner roundness (0..1)
            rotate: 0, // The rotation offset
            direction: 1, // 1: clockwise, -1: counterclockwise
            color: '#5C5C5C', // #rgb or #rrggbb or array of colors
            speed: 1, // Rounds per second
            trail: 60, // Afterglow percentage
            shadow: false, // Whether to render a shadow
            hwaccel: false, // Whether to use hardware acceleration
            className: 'spinner', // The CSS class to assign to the spinner
            zIndex: 2e9, // The z-index (defaults to 2000000000)
            top: 'auto', // Top position relative to parent in px
            left: 'auto' // Left position relative to parent in px
        },

        startWaiting: function () {
            this._spinView.appendTo(this._parentView);
            this._spinner.spin(this._spinView[0]);
        },
        stopWaiting: function () {
            this._spinner.stop();
            this._spinView.remove();
        }
    };

    return WaitView;

});