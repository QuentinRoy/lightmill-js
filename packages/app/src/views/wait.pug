.lightmill-base-app.lightmill-base-app-wait
  .lightmill-puck-wrapper
    .lightmill-puck.lightmill-puck-1
    .lightmill-puck.lightmill-puck-2
