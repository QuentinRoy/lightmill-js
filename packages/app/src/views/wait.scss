@import './common';

$puck-wrapper-size: 60px;
$puck-radius: 5px;
$puck-color: #5e5e5e;
$background-color: #d1d1d1;
$rotation-duration: 1s;
$animation-name: #{$app-class}-wait-spin;

.#{$app-class}-wait {
  background-color: $background-color;

  .#{$class-prefix}puck-wrapper {
    position: absolute;
    top: 50%;
    left: 50%;
    animation: $animation-name $rotation-duration infinite linear;
    margin-top: -$puck-wrapper-size / 2;
    margin-left: -$puck-wrapper-size / 2;
    width: $puck-wrapper-size;
    height: $puck-wrapper-size;

    .#{$class-prefix}puck {
      position: absolute;
      border-radius: $puck-radius;
      background-color: $puck-color;
      width: $puck-radius * 2;
      height: $puck-radius * 2;
    }

    .#{$class-prefix}puck-1 {
      top: 0;
      left: 0;
    }

    .#{$class-prefix}puck-2 {
      right: 0;
      bottom: 0;
    }
  }
}

@keyframes #{$animation-name} {
  0%   {transform: rotate(0deg)}
  100% {transform: rotate(360deg)}
}
