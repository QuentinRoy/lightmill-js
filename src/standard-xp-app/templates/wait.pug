.xp-app-base.xp-app-base-wait
  .puck-wrapper
    .puck
    
