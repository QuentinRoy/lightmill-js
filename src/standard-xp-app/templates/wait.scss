@import './common';

$puck-wrapper-size: 400px;
$puck-radius: 25px;
$puck-color: #5e5e5e;
$background-color: #d1d1d1;
$rotation-duration: 1s;
$animation-name: #{$app-class}-wait-spin;

.#{$app-class}-wait {
  background-color: $background-color;

  .puck-wrapper {
    position: absolute;
    top: 50%;
    left: 50%;
    margin-top: -$puck-wrapper-size / 2;
    margin-left: -$puck-wrapper-size / 2;
    width: $puck-wrapper-size;
    height: $puck-wrapper-size;

    .puck {
      position: absolute;
      top: 0;
      left: 0;
      animation: $animation-name $rotation-duration infinite linear;
      border-radius: $puck-radius;
      background-color: $puck-color;
      width: $puck-radius * 2;
      height: $puck-radius * 2;
    }
  }
}

@keyframes #{$animation-name} {
  0%   {transform: rotate(0deg)}
  100% {transform: rotate(360deg)}
}
