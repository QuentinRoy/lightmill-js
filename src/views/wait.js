import Spinner from 'spin';

const spinnerArgs = {
  lines: 11, // The number of lines to draw
  length: 20, // The length of each line
  width: 10, // The line thickness
  radius: 30, // The radius of the inner circle
  corners: 1, // Corner roundness (0..1)
  rotate: 0, // The rotation offset
  direction: 1, // 1: clockwise, -1: counterclockwise
  color: 'white', // #rgb or #rrggbb or array of colors
  speed: 1, // Rounds per second
  trail: 60, // Afterglow percentage
  shadow: false, // Whether to render a shadow
  hwaccel: false, // Whether to use hardware acceleration
  className: 'spinner', // The CSS class to assign to the spinner
  zIndex: 2e9, // The z-index (defaults to 2000000000)
  top: 'auto', // Top position relative to parent in px
  left: 'auto' // Left position relative to parent in px
};

export default class WaitView {
  constructor(parentNode) {
    this._parentNode = parentNode;
    this._spinDiv = document.createElement('div');
    Object.assign(this._spinDiv.style, {
      position: 'absolute',
      width: '100%',
      height: '100%',
      margin: 0,
      padding: 0,
      top: 0,
      left: 0
    });
    this._spinDiv.style.backgroundColor = 'black';
    this._spinner = new Spinner(spinnerArgs);
    this._isShown = false;
  }
  show() {
    if (!this.isShown) {
      this._parentNode.appendChild(this._spinDiv);
      this._spinner.spin(this._spinDiv);
      this._isShown = true;
    }
  }
  hide() {
    if (this.isShown) {
      this._spinner.stop();
      if (this._parentNode.contains(this._spinDiv)) {
        this._parentNode.removeChild(this._spinDiv);
      }
      this._isShown = false;
    }
  }
  get isShown() {
    return this._isShown;
  }
}
