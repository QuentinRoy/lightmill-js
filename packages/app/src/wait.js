/* eslint-disable no-param-reassign */

import waitTemplate from './views/wait.pug';
import './views/wait.scss';

export default (node, message) => {
  node.innerHTML = waitTemplate({ message });

  return {
    remove() {
      node.innerHTML = '';
    }
  };
};
