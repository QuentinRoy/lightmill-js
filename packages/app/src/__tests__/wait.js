/* globals document */

import wait from '../wait';

let node;
beforeEach(() => {
  node = document.createElement('div');
  node.innerHTML = '<div>Foo</div>';
});

describe('Waiting view', () => {
  it('can be shown', () => {
    wait(node, 'message');
    expect(node).toMatchSnapshot();
  });
  it('can be removed', () => {
    const { remove } = wait(node, 'message');
    remove();
    expect(node).toMatchSnapshot();
  });
});
