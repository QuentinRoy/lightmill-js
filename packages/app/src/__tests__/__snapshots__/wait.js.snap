// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Waiting view can be removed 1`] = `<div />`;

exports[`Waiting view can be shown 1`] = `
<div>
  <div
    class="lightmill-base-app lightmill-base-app-wait"
  >
    <div
      class="lightmill-puck-wrapper"
    >
      <div
        class="lightmill-puck lightmill-puck-1"
      />
      <div
        class="lightmill-puck lightmill-puck-2"
      />
    </div>
  </div>
</div>
`;
